kclvm/tools/src/query/override.rs

```diff
@@ -1,50 +1,242 @@
+use std::collections::{HashMap, HashSet};
+
+use anyhow::{anyhow, Result};
+
 use kclvm_ast::walker::MutSelfMutWalker;
 use kclvm_ast::{ast, walk_if_mut};
-use kclvm_parser::parse_expr;
+use kclvm_parser::{parse_expr, parse_file};
+
+use crate::printer::print_ast_module;
+
+/// Override and rewrite a file with override spec.
+///
+/// # Parameters
+///
+/// `file`: [&str]
+///     The File that need to be overridden
+///
+/// `specs`: &\[[String]\]
+///     List of specs that need to be overridden.
+///     Each spec string satisfies the form: <pkgpath>:<field_path>=<filed_value> or <pkgpath>:<field_path>-
+///     When the pkgpath is '__main__', it can be omitted.
+///
+/// `import_paths`: &\[[String]\]
+///     List of import paths that are need to be added.
+///
+/// # Returns
+///
+/// result: [Result<bool>]
+///     Whether override is successful
+///
+/// # Examples
+///
+/// ```no_run
+/// use kclvm_tools::query::override_file;
+///
+/// let result = override_file(
+///     "test.k",
+///     &["alice.age=18".to_string()],
+///     &[]
+/// ).unwrap();
+/// ```
+///
+/// - test.k (before override)
+///
+/// ```kcl
+/// schema Person:
+///     age: int
+///
+/// alice = Person {
+///     age = 10
+/// }
+/// ```
+///
+/// - test.k (after override)
+///
+/// ```kcl
+/// schema Person:
+///     age: int
+///
+/// alice = Person {
+///     age = 18
+/// }
+/// ```
+pub fn override_file(file: &str, specs: &[String], import_paths: &[String]) -> Result<bool> {
+    let overrides = specs
+        .iter()
+        .map(|s| parse_override_spec(s))
+        .filter_map(Result::ok)
+        .collect::<Vec<ast::OverrideSpec>>();
+    let mut module = match parse_file(file, None) {
+        Ok(module) => module,
+        Err(msg) => return Err(anyhow!("{}", msg)),
+    };
+    let mut result = false;
+    for o in &overrides {
+        if apply_override_on_module(&mut module, o, import_paths)? {
+            result = true;
+        }
+    }
+    if result {
+        let code_str = print_ast_module(&module);
+        std::fs::write(file, &code_str)?
+    }
+    Ok(result)
+}
+
+/// Override spec string to override structure
+///
+/// # Example
+///
+/// ```
+/// use kclvm_tools::query::parse_override_spec;
+/// use kclvm_ast::ast;
+///
+/// let spec = parse_override_spec("alice.age=10").unwrap();
+/// assert_eq!(spec, ast::OverrideSpec {
+///     pkgpath: "".to_string(),
+///     field_path: "alice.age".to_string(),
+///     field_value: "10".to_string(),
+///     action: ast::OverrideAction::CreateOrUpdate,
+/// });
+/// ```
+pub fn parse_override_spec(spec: &str) -> Result<ast::OverrideSpec> {
+    let err = Err(invalid_spec_error(spec));
```

Comment:
move the declaration to where it is used

Reply:
> move the declaration to where it is used

move it to the mod.rs

Reply:
move it to `override.rs` and set it to private.